plugins/scaffolder-node/src/actions/types.ts

```diff
@@ -63,14 +63,24 @@ export type ActionContext<TInput extends JsonObject> = {
 };
 
 /** @public */
-export type TemplateAction<TInput extends JsonObject> = {
+export type TemplateAction<
+  TParams = {},
+  TInputSchema extends Schema | z.ZodType = {},
+  TOutputSchema extends Schema | z.ZodType = {},
+> = {
   id: string;
   description?: string;
   examples?: { description: string; example: string }[];
   supportsDryRun?: boolean;
   schema?: {
-    input?: Schema;
-    output?: Schema;
+    input?: TInputSchema;
+    output?: TOutputSchema;
```

Comment:
Hmm, should we not keep this as always being JSONSchema? Just that we allow a zod schema as the input to `createTemplateAction`? Seems like we end up leaving a lot to be figured out in the usage of the actions now instead. That way we should be able to infer the types in `createTemplateAction` and avoid leaking the zod types too, similar to [makeFieldSchemaFromZod](https://github.com/backstage/backstage/blob/da25293910f208d00296af7c7de497a8f8ff7278/plugins/scaffolder/src/components/fields/utils.ts#L40)

Reply:
Hmm maybe we can get together and pair a bit on this then. Maybe there's a better way to do this instead of the types being polluted and we can do the `zod-to-jsonschema` in the `createTemplateAction` instead? Is that what you mean?